File: macos/Sources/LumaMDCore/MathMLNode.swift
import Foundation

indirect enum MathMLNode {
    case row([MathMLNode])
    case identifier(String)
    case number(String)
    case operatorSymbol(String, limits: Bool)
    case text(String)
    case textLiteral(String)
    case fraction(MathMLNode, MathMLNode)
    case squareRoot(MathMLNode)
    case root(MathMLNode, MathMLNode)
    case scripts(MathMLNode, subscript: MathMLNode?, superscript: MathMLNode?)
    case fenced(left: String?, body: MathMLNode, right: String?)
    case accent(MathMLNode, String)
    case style(MathMLNode, variant: String)
    case operatorName(String)
    case table(rows: [[MathMLNode]], alignment: String?, left: String?, right: String?)

    func render() -> String {
        switch self {
        case .row(let nodes):
            let content = nodes.map { $0.render() }.joined()
            if nodes.count == 1 { return content }
            return "<mrow>\(content)</mrow>"
        case .identifier(let value):
            return "<mi>\(escape(value))</mi>"
        case .number(let value):
            return "<mn>\(escape(value))</mn>"
        case .operatorSymbol(let value, _):
            return "<mo>\(escape(value))</mo>"
        case .text(let value):
            return "<mtext>\(escape(value))</mtext>"
        case .textLiteral(let value):
            return renderTextLiteral(value)
        case .fraction(let numerator, let denominator):
            return "<mfrac>\(numerator.render())\(denominator.render())</mfrac>"
        case .squareRoot(let body):
            return "<msqrt>\(body.render())</msqrt>"
        case .root(let body, let index):
            return "<mroot>\(body.render())\(index.render())</mroot>"
        case .scripts(let base, let lower, let upper):
            return renderScripts(base: base, lower: lower, upper: upper)
        case .fenced(let left, let body, let right):
            return "<mrow>\(fence(left))\(body.render())\(fence(right))</mrow>"
        case .accent(let body, let accent):
            return "<mover accent=\"true\">\(body.render())<mo>\(escape(accent))</mo></mover>"
        case .style(let body, let variant):
            return "<mstyle mathvariant=\"\(variant)\">\(body.render())</mstyle>"
        case .operatorName(let value):
            return "<mo>\(escape(value))</mo>"
        case .table(let rows, let alignment, let left, let right):
            let attribute = alignment.map { " columnalign=\"\($0)\"" } ?? ""
            let body = rows.map { row in
                "<mtr>" + row.map { "<mtd>\($0.render())</mtd>" }.joined() + "</mtr>"
            }.joined()
            let table = "<mtable\(attribute)>\(body)</mtable>"
            guard left != nil || right != nil else { return table }
            return "<mrow>\(fence(left))\(table)\(fence(right))</mrow>"
        }
    }

    private var usesLimits: Bool {
        if case .operatorSymbol(_, let limits) = self { return limits }
        return false
    }

    private func renderTextLiteral(_ value: String) -> String {
        var nodes = [String]()
        var text = ""
        for character in value {
            if character == " " {
                if !text.isEmpty {
                    nodes.append("<mtext>\(escape(text))</mtext>")
                    text = ""
                }
                nodes.append("<mspace width=\"0.28em\"></mspace>")
            } else {
                text.append(character)
            }
        }
        if !text.isEmpty {
            nodes.append("<mtext>\(escape(text))</mtext>")
        }
        return nodes.count == 1 ? nodes[0] : "<mrow>\(nodes.joined())</mrow>"
    }

    private func renderScripts(base: MathMLNode, lower: MathMLNode?, upper: MathMLNode?) -> String {
        switch (lower, upper) {
        case let (.some(lower), .some(upper)) where base.usesLimits:
            return "<munderover>\(base.render())\(lower.render())\(upper.render())</munderover>"
        case let (.some(lower), .none) where base.usesLimits:
            return "<munder>\(base.render())\(lower.render())</munder>"
        case let (.none, .some(upper)) where base.usesLimits:
            return "<mover>\(base.render())\(upper.render())</mover>"
        case let (.some(lower), .some(upper)):
            return "<msubsup>\(base.render())\(lower.render())\(upper.render())</msubsup>"
        case let (.some(lower), .none):
            return "<msub>\(base.render())\(lower.render())</msub>"
        case let (.none, .some(upper)):
            return "<msup>\(base.render())\(upper.render())</msup>"
        case (.none, .none):
            return base.render()
        }
    }

    private func fence(_ delimiter: String?) -> String {
        guard let delimiter else { return "" }
        return "<mo fence=\"true\" stretchy=\"true\">\(escape(delimiter))</mo>"
    }

    private func escape(_ value: String) -> String {
        InlineMarkdown.escape(value)
    }
}

File: app/src/main/java/dev/lumamd/viewer/core/MathMLNode.java
package dev.lumamd.viewer.core;

import java.util.List;

final class MathMLNode {
    private final String html;
    private final boolean limits;

    private MathMLNode(String html, boolean limits) {
        this.html = html;
        this.limits = limits;
    }

    static MathMLNode row(List<MathMLNode> nodes) {
        StringBuilder html = new StringBuilder("<mrow>");
        for (MathMLNode node : nodes) {
            html.append(node.render());
        }
        return new MathMLNode(html.append("</mrow>").toString(), false);
    }

    static MathMLNode identifier(String value) {
        return element("mi", value, false);
    }

    static MathMLNode number(String value) {
        return element("mn", value, false);
    }

    static MathMLNode operator(String value) {
        return element("mo", value, false);
    }

    static MathMLNode limitOperator(String value) {
        return element("mo", value, true);
    }

    static MathMLNode text(String value) {
        return element("mtext", value, false);
    }

    static MathMLNode spacedText(String value) {
        List<MathMLNode> nodes = new java.util.ArrayList<MathMLNode>();
        StringBuilder text = new StringBuilder();
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if (character == ' ') {
                if (text.length() > 0) {
                    nodes.add(text(text.toString()));
                    text.setLength(0);
                }
                nodes.add(new MathMLNode("<mspace width=\"0.28em\"></mspace>", false));
            } else {
                text.append(character);
            }
        }
        if (text.length() > 0) {
            nodes.add(text(text.toString()));
        }
        return nodes.size() == 1 ? nodes.get(0) : row(nodes);
    }

    static MathMLNode fraction(MathMLNode numerator, MathMLNode denominator) {
        return composite("mfrac", numerator, denominator);
    }

    static MathMLNode squareRoot(MathMLNode content) {
        return composite("msqrt", content);
    }

    static MathMLNode indexedRoot(MathMLNode content, MathMLNode index) {
        return composite("mroot", content, index);
    }

    static MathMLNode subscript(MathMLNode base, MathMLNode script) {
        return composite("msub", base, script);
    }

    static MathMLNode superscript(MathMLNode base, MathMLNode script) {
        return composite("msup", base, script);
    }

    static MathMLNode subSuperscript(
            MathMLNode base,
            MathMLNode subscript,
            MathMLNode superscript) {
        return composite("msubsup", base, subscript, superscript);
    }

    static MathMLNode under(MathMLNode base, MathMLNode script) {
        return composite("munder", base, script);
    }

    static MathMLNode over(MathMLNode base, MathMLNode script) {
        return composite("mover", base, script);
    }

    static MathMLNode underOver(
            MathMLNode base,
            MathMLNode underscript,
            MathMLNode overscript) {
        return composite("munderover", base, underscript, overscript);
    }

    static MathMLNode fenced(String open, String close, MathMLNode content) {
        return fenced(open, close, content, null);
    }

    static MathMLNode fenced(
            String open,
            String close,
            MathMLNode content,
            String minimumSize) {
        StringBuilder html = new StringBuilder("<mrow>");
        if (!open.isEmpty()) {
            html.append("<mo fence=\"true\" stretchy=\"true\"");
            if (minimumSize != null) {
                html.append(" minsize=\"")
                        .append(InlineMarkdown.escape(minimumSize))
                        .append("\"");
            }
            html.append(">")
                    .append(InlineMarkdown.escape(open))
                    .append("</mo>");
        }
        html.append(content.render());
        if (!close.isEmpty()) {
            html.append("<mo fence=\"true\" stretchy=\"true\"");
            if (minimumSize != null) {
                html.append(" minsize=\"")
                        .append(InlineMarkdown.escape(minimumSize))
                        .append("\"");
            }
            html.append(">")
                    .append(InlineMarkdown.escape(close))
                    .append("</mo>");
        }
        return new MathMLNode(html.append("</mrow>").toString(), false);
    }

    static MathMLNode environmentFenced(
            String open,
            String close,
            MathMLNode content,
            int rowCount) {
        return new MathMLNode(
                "<mrow>" + fenceTable(open, rowCount, true)
                        + content.render()
                        + fenceTable(close, rowCount, false)
                        + "</mrow>",
                false);
    }

    static MathMLNode accent(MathMLNode base, String mark) {
        return new MathMLNode(
                "<mover accent=\"true\">" + base.render() + operator(mark).render() + "</mover>",
                false);
    }

    static MathMLNode style(MathMLNode content, String variant) {
        return new MathMLNode(
                "<mstyle mathvariant=\"" + InlineMarkdown.escape(variant) + "\">"
                        + content.render() + "</mstyle>",
                false);
    }

    static MathMLNode table(List<List<MathMLNode>> rows) {
        StringBuilder html = new StringBuilder("<mtable>");
        for (List<MathMLNode> row : rows) {
            html.append("<mtr>");
            for (MathMLNode cell : row) {
                html.append("<mtd>").append(cell.render()).append("</mtd>");
            }
            html.append("</mtr>");
        }
        return new MathMLNode(html.append("</mtable>").toString(), false);
    }

    boolean hasLimits() {
        return limits;
    }

    String render() {
        return html;
    }

    private static MathMLNode element(String tag, String value, boolean limits) {
        return new MathMLNode(
                "<" + tag + ">" + InlineMarkdown.escape(value) + "</" + tag + ">",
                limits);
    }

    private static MathMLNode composite(String tag, MathMLNode... nodes) {
        StringBuilder html = new StringBuilder("<").append(tag).append('>');
        for (MathMLNode node : nodes) {
            html.append(node.render());
        }
        return new MathMLNode(html.append("</").append(tag).append('>').toString(), false);
    }

    private static String fenceTable(String delimiter, int rowCount, boolean opening) {
        if (delimiter.isEmpty()) {
            return "";
        }
        String top;
        String middle;
        String bottom;
        boolean mirrored = ")".equals(delimiter);
        if ("(".equals(delimiter) || mirrored) {
            top = "⎛";
            middle = "⎜";
            bottom = "⎝";
        } else if ("{".equals(delimiter)) {
            top = opening ? "⎧" : "⎫";
            middle = opening ? "⎨" : "⎬";
            bottom = opening ? "⎩" : "⎭";
        } else {
            return operator(delimiter).render();
        }

        int safeRows = Math.max(2, rowCount);
        StringBuilder html = new StringBuilder(
                "<mtable class=\"math-fence\" rowspacing=\"0\" columnspacing=\"0\">");
        for (int row = 0; row < safeRows; row++) {
            String piece;
            if (row == 0) {
                piece = top;
            } else if (row == safeRows - 1) {
                piece = bottom;
            } else if ("{".equals(delimiter) && row == safeRows / 2) {
                piece = middle;
            } else {
                piece = "(".equals(delimiter) || mirrored ? middle : "⎪";
            }
            html.append("<mtr><mtd><mo")
                    .append(mirrored ? " class=\"math-fence-mirror\"" : "")
                    .append(" fence=\"true\">")
                    .append(piece)
                    .append("</mo></mtd></mtr>");
        }
        return html.append("</mtable>").toString();
    }
}
